### PR TITLE
IMap table metadata resolution (#16984)

### LINKED ISSUE #16984: SQL: Implement IMap table metadata resolution
We need to implement the metadata source for `IMap`. The implementation should expose registered IMap instances as tables in the following way:
1. Every map is exposed as a table with the same name
1. All IMap-backed tables are located in the predefined schema `partitioned`
1. This schema is added to the default search paths, thus allowing references to objects without an explicit schema name. That is, both `FROM myMap` and `FROM partitioned.myMap` are valid
1. Fields are derived through sampling: we get the very first local map entry and analyze its getters and public fields to construct the list of fields
1. Key and value could be accessed via `__key` and `this` respectively, which is compliant with existing Predicate API
1. '__key' is always exposed as a normal field to maintain uniqueness
1. `this` is exposed as a normal field if it is of a basic type, or as a hidden field if it is a composite type

### PR DESCRIPTION
**Pull request by devozerov**
This PR introduces automatic metadata resolution for `IMap`s. Implementation notes:
- Every `IMap` is represented as a table in the schema `partitioned` with the same name
- There is a search path `hazelcast.partitioned` that allows us to refer to the table both as `SELECT ... FROM table` and `SELECT ... FROM partitioned.table`
- To resolve the fields, we analyze the first local key-value pair and extract public getters and public fields from it. Getters take precedence over fields in case of conflict. Likewise, child class members take precedence over parent class members. If there is no local entry on the member, an exception will be thrown lazily when the map is accessed during query evaluation 
- Supported in-memory formats are `OBJECT` and `BINARY`. `NATIVE` will throw an exception since we do not support HD yet
- JSON type is not supported
- Table statistics are calculated by counting local owning entries and then multiplying them to the member count
- Currently, resolution happens eagerly on every SQL request. We may want to implement some caching in the future, but this is not a concern of the current implementation

Closes #16984

### CODE REVIEW

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/954/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/954/Hazelcast-pr-builder-954.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   DataSerializableConventionsTest.test_dataSerializableClasses_areIdentifiedDataSerializable:121 There are 1 classes which are DataSerializable, not @BinaryInterface-annotated and are not IdentifiedDataSerializable.
[INFO] 
[ERROR] Tests run: 3855, Failures: 1, Errors: 0, Skipped: 14
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Comment by devozerov** [comment]
run-lab-run

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/955/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/955/Hazelcast-pr-builder-955.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ClientCacheEventJournalBasicTest>AbstractEventJournalBasicTest.allowReadingWithFutureSeq:512->HazelcastTestSupport.assertOpenEventually:1108->HazelcastTestSupport.assertOpenEventually:1123 CountDownLatch failed to complete within 30 seconds, count left: 1
[ERROR]   ClientMapEventJournalBasicTest>AbstractEventJournalBasicTest.allowReadingWithFutureSeq:512->HazelcastTestSupport.assertOpenEventually:1108->HazelcastTestSupport.assertOpenEventually:1123 CountDownLatch failed to complete within 30 seconds, count left: 1
[INFO] 
[ERROR] Tests run: 27748, Failures: 2, Errors: 0, Skipped: 1001
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

File: hazelcast/src/main/java/com/hazelcast/sql/impl/extract/QueryPath.java [hunk 1]
```diff
@@ -94,4 +104,59 @@ private static boolean isEmpty(String path) {
...
+        if (key != path1.key) {
+            return false;
+        }
+
+        return path != null ? path.equals(path1.path) : path1.path == null;
+    }
+
+    @Override
+    public int hashCode() {
+        int result = (key ? 1 : 0);
```
Comment (petrpleshachkov) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/extract/QueryPath.java:151]: Minor : parentheses are not needed

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/PartitionedMapTableResolver.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+                    fields,
+                    new ConstantTableStatistics(estimatedRowCount),
+                    keyMetadata.getDescriptor(),
+                    valueMetadata.getDescriptor()
+                );
+            }
+
+            return emptyMap(name);
+        } catch (QueryException e) {
+            throw e;
```
Comment (petrpleshachkov) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/PartitionedMapTableResolver.java:170]: `return new PartitionedMapTable(mapName, e);` ?

File: hazelcast/src/main/java/com/hazelcast/sql/impl/extract/QueryPath.java (hunk 1 above)
Comment (devozerov) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/extract/QueryPath.java:151]: Fixed.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/PartitionedMapTableResolver.java (hunk 2 above)
Comment (devozerov) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/PartitionedMapTableResolver.java]: Refactored.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadata.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        this.descriptor = descriptor;
+        this.fields = fields;
+    }
+
+    public QueryTargetDescriptor getDescriptor() {
+        return descriptor;
+    }
+
+    @SuppressWarnings("checkstyle:IllegalType")
+    public LinkedHashMap<String, TableField> getFields() {
```
Comment (petrpleshachkov) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadata.java:42]: Do we have to expose here `LinkedHashMap`? Maybe `Map` is enough?

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/AbstractMapTable.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+        return super.getStatistics();
+    }
+
+    public QueryTargetDescriptor getKeyDescriptor() {
+        return keyDescriptor;
+    }
+
+    public QueryTargetDescriptor getValueDescriptor() {
+        return valueDescriptor;
```
Comment (taburet) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/AbstractMapTable.java:87]: checkException()?

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+        InternalSerializationService ss,
+        Object target,
+        boolean binary,
+        boolean key
+    ) {
+        try {
+            if (target instanceof Data) {
+                Data data = (Data) target;
+
+                if (data.isPortable() && binary) {
```
Comment (taburet) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java:73]: Is there any specific reason for using portable style resolution only for binary? If user would change the storage format, we may end up with a different schema.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadata.java (hunk 3 above)
Comment (devozerov) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadata.java:42]: Order of fields is important for us, `LinkedHashMap` reflects that.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/AbstractMapTable.java (hunk 4 above)
Comment (devozerov) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/AbstractMapTable.java:87]: `checkException()` is needed to let optimizer know that the table exists, but is not functional. Descriptors are used after the optimization process, so an exception check is not needed here. To summarize, exception check is only needed in the overridden methods.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java (hunk 5 above)
Comment (devozerov) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java:73]: The idea is as follows:
1. Background: in HZ key is always stored as Data, value is stored as either object or key. Portable object exposes different fields depending on whether we process it as object or as Data
1. If the format is OBJECT, we always deserialize to real objects
2. If the format is BINARY, we do not deserialize portables
This works almost in the same way as in Predicate API with one difference: if both key and value as Portables and storage format is OBJECT, then we extract Java fields form both key and value, while in Predicate API we extract portable fields from key, and Java fields from value. 
Our approach looks more consistent to me - user asked us to store key-value pairs In a form of Java objects, and we expose fields of Java objects from both key and value.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/AbstractMapTable.java (hunk 4 above)
Comment (taburet) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/AbstractMapTable.java:87]: assert exception == null?

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java (hunk 5 above)
Comment (taburet) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java:73]: Another option is to resolve the schema for OBJECT format using the portable definition, but read the values from the Java object itself (I think it was done this way for performance reasons). But that might produce even more weird behavior, looks like we can't do anything about it.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/AbstractMapTable.java (hunk 4 above)
Comment (devozerov) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/AbstractMapTable.java:87]: Maybe, but again - exceptions are needed by Calcite integration, that works with `Table` interface, and never touch extension methods.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java (hunk 5 above)
Comment (devozerov) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java:73]: The problem is that portable definition may differ from Java fields.

**Pull request merged**